Add initial infrastructure for merkle tree storage

dartdart26 opened the pull request:
In order to support parallel implementations of both existing (direct
key value) and future (merkle tree based) storage systems, provide an
initial infrastructure for the implementation of DBKeyManipulator and
DBAdapter for merkle tree based storage.

This infrastructure includes simplified key serialization routines for
the newly introduced merkle tree based keys. Additionally, put common
DBKeyManipulator and DBAdapter functionalities in respective base
classes to facilitate reuse.

Introduce an initial version of the DBAdapter::Reader class that
implements sparse_merkle::IDBReader in order to bridge the merkle tree
to the storage engine. Further commits will provide full implementation.

Finally, add sparse merkle key manipulation and initial DB adapter
tests.

dartdart26 commented:
Added DB adapter functionality and tests:

- `getKeyByReadVersion()`
- `getLatestBlock()`

dartdart26 commented:
Added code to convert updates returned by the merkle tree to key/value pairs suitable for the DB.

andrewjstone commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+}
+
+BlockId DBKeyManipulator::extractBlockIdFromKey(const concordUtils::Key &key) {
+  const auto offset = key.length() - sizeof(BlockId);
```
This works, but any change to the LeafKey structure or addition of another key type will break it. Maybe add a comment that this only works for a serialized `LeafKey`? Later on we should have round trip tests so we can catch breakages in deserialization.

dartdart26 replied:
Yeah, that's a good point. I only kept it to match up with the existing interface. As of now, it should work for block and leaf keys. I think we should not have that function in the future (or at least not in that form). I will add the comment for now.

andrewjstone commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+    actualVersion = DBKeyManipulator::extractBlockIdFromKey(foundKey);
+  }
+
+  return Status::OK();
```
I really think we misuse status responses here. I realize that this likely matches up with existing behavior, but it's probably worth fixing in the future to return `Status::NotFound` when a key is not found rather than `Status::OK`. Then we'd have a more explicit check rather than checking if the key is empty.

dartdart26 replied:
Yes, totally agree.

dartdart26 commented on storage/include/memorydb/client.h:
```diff
@@ -58,9 +58,12 @@ class ClientIterator : public concord::storage::IDBClient::IDBClientIterator {
 // single thread.
+//
+// The default KeyComparator provides lexicographical ordering. If 'a' is shorter than 'b' and they match up to the
+// length of 'a', then 'a' is considered to preceed 'b'.
```
Fixed.

dartdart26 commented on util/include/endianness.hpp:
```diff
@@ -0,0 +1,41 @@
+    v = htonl(v);
+  } else if constexpr (sizeof(v) == 8) {
+    const auto high = htonl(v >> 32);
+    const auto low = htonl(v & std::numeric_limits<std::uint32_t>::max());
```
What a great catch! It made me thinking that a narrowing unsigned cast is well-defined and we can just pass `v`.

dartdart26 commented on storage/test/merkleTreeAdapter_test.cpp:
```diff
@@ -0,0 +1,269 @@
+namespace {
+
+template <typename E>
+std::string eToS(E e) {
```
Done.

dartdart26 commented on storage/test/merkleTreeAdapter_test.cpp:
```diff
@@ -0,0 +1,269 @@
+
+template <typename T>
+std::string toS(T v) {
+  v = concordUtils::hostToNet(v);
```
It is like that, because we use the string as a byte buffer. I have called the function `serializeIntegral`.

andrewjstone replied:
Ah yeah, I meant to remove this comment when I went back and reviewed the creation code.

dartdart26 commented on storage/test/merkleTreeAdapter_test.cpp:
```diff
@@ -0,0 +1,269 @@
+TEST(key_manipulator, st_res_page_dynamic_key) {
+  const auto key = DBKeyManipulator::generateSTReservedPageDynamicKey(defaultPageId, defaultChkpt);
+  const auto expected =
+      toSl(eToS(EDBKeyType::BFT) + eToS(EBFTType::STReserverdPageDynamic) + toS(defaultPageId) + toS(defaultChkpt));
```
Fixed.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+namespace {
+
+// Specifies the key type. Used when serializing the stale node index.
+enum class KeySerializationType : std::uint8_t {
```
Renaming to `StaleKeyType`.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+};
+
+// Specifies the child type. Used when serializing BatchedInternalNode objects.
+enum class ChildSerializationType : std::uint8_t { Leaf, Internal };
```
Done.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+
+std::string serializeImp(KeySerializationType type) { return std::string{toChar(type)}; }
+
+// std::string serializeImp(ChildSerializationType type) { return std::string{toChar(type)}; }
```
Done.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+    }
+  }
+
+  // Serialize by putting a 32 bit bitmask specifying which indixes are set in the children array. Then, put a type
```
Done.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+
+std::string serializeImp(EDBKeyType type) { return std::string{toChar(type)}; }
+
+std::string serializeImp(EKeyType type) { return std::string{toChar(EDBKeyType::Key), toChar(type)}; }
```
Renamed to `EKeySubtype` and `EBFTSubtype` .

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+std::string serializeImp(const std::vector<std::uint8_t> &v) { return std::string{std::cbegin(v), std::cend(v)}; }
+
+std::string serializeImp(const NibblePath &path) {
+  return serializeImp(static_cast<std::uint8_t>(path.length())) + serializeImp(path.data());
```
I think you are right.

My thinking was:

- having a simple interface that has an input and output is cleaner
- std::string has a small string optimisation (maybe around 20-24 bytes) and also a strategy for allocation that allocates more memory (e.g. twice the needed size)
- operator+ has overloads with rvalue references and we typically chain temporaries

Basically, based on above, it shouldn't allocate often (haven't measured it, though) and it can sometimes even avoid allocation.

I agree on the benchmarks and measurements, let's investigate that later.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+}
+
+// Converts the updates as returned by the merkle tree to key/value pairs suitable for the DB.
+SetOfKeyValuePairs batchToDBUpdates(const Tree::UpdateBatch &batch) {
```
I am planning to provide that when implementing `addBlock()` and `getBlockById()`.

dartdart26 commented on storage/src/merkle_tree_db_adapter.cpp:
```diff
@@ -0,0 +1,323 @@
+
+Sliver DBKeyManipulator::genDataDbKey(const Sliver &key, BlockId version) {
+  auto hasher = Hasher{};
+  return serialize(EKeyType::Leaf, LeafKey{hasher.hash(key.data(), key.length()), version});
```
Good catch, doing that.

dartdart26 commented:
@andrewjstone Thanks for the thorough review! I have addressed your comments. Additionally, I have added a comment on the ordering of keys such that we can support the requested operations. You can check it in `db_types.h`. Pushed based on above.

andrewjstone commented:
Ship it!
